core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -418,6 +424,15 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf,
     cachedSerializedStatuses.contains(shuffleId) || mapStatuses.contains(shuffleId)
   }
 
+  /** Get the epoch for map output for a shuffle, if it is available */
+  def getEpochForMapOutput(shuffleId: Int, mapId: Int): Option[Long] = {
+    val arrayOpt = mapStatuses.get(shuffleId)
+    if (arrayOpt.isDefined && arrayOpt.get != null && arrayOpt.get(mapId) != null) {
+       return Some(epochForMapStatus.get(shuffleId).get(mapId))
+    }
+    None
+  }
```

Comment:
First, `arrayOpt.get != null` isn't necessary since we don't put `null` values into `mapStatuses`. Second, `epochForMapStatus.get(shuffleId).get` is the same as `epochForMapStatus(shuffleId)`. Third, I don't like all the explicit `get`s,`null` checks and the unnecessary non-local `return`. To my mind, this is better:
``` scala
  def getEpochForMapOutput(shuffleId: Int, mapId: Int): Option[Long] = {
    for {
      mapStatus <- mapStatuses.get(shuffleId).flatMap { mapStatusArray =>
        Option(mapStatusArray(mapId))
      }
    } yield epochForMapStatus(shuffleId)(mapId)
  }
```